### PR TITLE
Part units

### PR DESCRIPTION
**Pull request by SchrodingersGat**
Closes https://github.com/inventree/InvenTree/issues/4829

This PR changes the "units" field for a part to be a physical unit (e.g. "m")

Additionally, supplier parts can be specified in *compatible* units (e.g "feet"), and converted to base units as appropriate

### TODO

- [x] Create custom units for "pieces" / "each" / etc
- [x] Unit testing for new custom units
- [x] Replace all references to old `pack_size` variable
- [x] Fix existing unit tests
- [x] Add unit tests for converting between unit types (i.e. "10 feet" -> "m")
- [x] Add unit tests to ensure supplier part pricing is correctly converted to base units
- [x] Add background task to ensure that supplier part units are updated when the parent part changes
- [x] Documentation
- [x] Add "native_pack_units" field to the SupplierPart model (which is calculated in base units, for queryset annotations)
- [x] Data migration for demo dataset - https://github.com/inventree/demo-dataset/pull/46
- [x] Data migration for converting exiting "part.units" to physical units
- [x] Unit test for above data migration
- [x] Data migration for converting any existing "pack_size" fields to new units
- [x] Unit test for above data migration
- [x] Ensure that existing forms which use supplier part quantity still work as expected
- [x] Ensure that pricing calculations are correctly considered when using pack_quantity

### CODE REVIEW

**Review by matmair — APPROVED** [review]
Would it make sense to introduce a unit 'each' or 'piece'/'pc' for a quantity of 1? I have seen that in other PLMs/ERPs

**Comment by SchrodingersGat** [comment]
> Would it make sense to introduce a unit 'each' or 'piece'/'pc' for a quantity of 1? I have seen that in other PLMs/ERPs

Yep, that makes sense. There's an easy entry point as I've already defined a hook for creating a custom UnitRegistry

File: InvenTree/company/models.py [hunk 1]
```diff
@@ -561,14 +601,26 @@ def save(self, *args, **kwargs):
...
+    pack_quantity_native = RoundingDecimalField(
+        max_digits=20, decimal_places=10, default=1,
+        null=True,
     )
 
+    def base_quantity(self, quantity=1) -> Decimal:
+        """Calculate the base unit quantiy for a given quantity."""
+
+        q = Decimal(quantity) * Decimal(self.pack_quantity_native)
+        q = round(q, 10).normalize()
```
Comment (matmair) [inline InvenTree/company/models.py:620]: Question: Should this rounding be noted in the docs?

**Review by matmair — APPROVED** [review]
On a general note: Should we consider squashing migrations with the next major release? These migrations that require 3 steps are culminating.

**Comment by matmair** [comment]
@SchrodingersGat have you seen that there are tests that are not run in the coverage?

**Comment by SchrodingersGat** [comment]
No, which tests are not runnign?

**Comment by matmair** [comment]
The migration tests do not seem to run fully anymore, coverage changed [here](https://coveralls.io/builds/60316457/source?filename=InvenTree%2Fpart%2Fmigrations%2F0109_auto_20230517_1048.py)

**Comment by SchrodingersGat** [comment]
> The migration tests do not seem to run fully anymore, coverage changed [here](https://coveralls.io/builds/60316457/source?filename=InvenTree%2Fpart%2Fmigrations%2F0109_auto_20230517_1048.py)

I think it should be fine

**Pull request merged**